### ISSUE TITLE
podman: single pull for parallel run

### ISSUE THREAD

**Issue by muayyad-alsadi** [containers/libpod#2551]
/kind feature

when doing `podman run` in parallel, it would trigger multiple pulls and download the same image multiple times

```
$ podman rmi busybox
d8233ab899d419c58cf3634c0df54ff5d8acc28f8173f09c21df4a07229e1205
$ for i in `seq 10`; do ( podman run -d --name b$i busybox:latest busybox httpd -f -p 80 )& done; wait
```

![screenshot from 2019-03-06 08-52-03](https://user-images.githubusercontent.com/1312683/53861611-50fbf780-3fed-11e9-8091-f9f0c0e286c3.png)

**Comment by vrothberg** [comment]
Thanks for opening the issue, @muayyad-alsadi.

As soon as any of the blobs or the config is in the local storage, Podman will not attempt to download it but reuse the existing data. This means that a given layer/config will only be downloaded in parallel if it's not yet in the storage.

Due to Podman's daemon-less architecture, I don't think we can avoid this "race" but we can try adding the "already present" case to the progress bars.

When running with `--log-level=debug`, you should see logs similar to:
```
DEBU[0004] Skipping blob sha256:6c40cc604d8e4c121adcb6b0bfe8bb038815c350980090e74aa5a6423f8f82c0 (already present):
```

**Comment by muayyad-alsadi** [comment]
> I don't think we can avoid this "race" 

we can use a lock file, I've done that even in my shell scripts.

```bash
set -x
LOCKFILE="/tmp/mytask.tmp"
[ "${FLOCKER}" != "$0" ] && exec env FLOCKER="$0" flock -nx "$LOCKFILE" "$0" "$@" || :

do_something_here
```

a good enough lock per image can be done on any file for that specific name or image id like the manifest or pull.lock inside that directory

```
~/.local/share/containers/storage/vfs-images/288...c5a/manifest
~/.local/share/containers/storage/vfs-images/288...c5a/pull.lock
```

beside lock by id, a lock per name is a good idea to make sure lock is done before pulling meta data and resolving name to id (busybox:latest from which registery and which version)

once the lock is released all other pull operation would give "skipping: blob already exists" which is the happy scenario

also lock per-layer id is also good.

**Comment by vrothberg** [comment]
> > I don't think we can avoid this "race"
> 
> we can use a lock file, I've done that even in my shell scripts.

Let me rephrase: I don't think we can avoid this race without major rewrites of the code and we have to consider the costs of locks. The devil lies in the details.

A big lock as shown in the bashscript is certainly a performance killer, so we had to make it more specific. If we made it specific to the image/ID, we would miss that two images can share the same base image(s). Creating locks on the granularity of layers would be even more expensive as more and more locks had to be created, locked and unlocked (all of them syscalls).

On the other hand, there are more possible destinations than container-storage (e.g., a directory, oci, docker-archive, docker daemon, etc.) and we could also *push* to all of them. Storing the locks in the storage would hence not suffice as we had to make it transport-generic and specific to the (image, source, destination) triple.

Some of those transports are immutable (e.g., a docker registry) while others are not, for instance, the directory transport. This means that some operations are safe and others will always lead to a potential data corruption - independent of serializing processes with locks or not. Then, there is the environment of the copy operation which includes if we're running as root or not, which storage driver we're using etc. The containers/image library does not support that (yet?).

In this context, we face a (rather rare?) case where multiple processes pull the same image. I completely understand that potentially downloading the same data multiple times is undesireable. However, it does not lead to a bug and in some cases, even when the execution times overlap, the data is already present and can be reused.

My conclusions is that it's a situation we have to live with but maybe others have a different opinion. @mtrmac @rhatdan @mheon @baude WDYT?

**Comment by muayyad-alsadi** [comment]
> consider the costs of locks.

no matter how many nano seconds the lock cost, it's not as big as downloading openjdk image twice then discard it because it's already downloaded.

> Creating locks on the granularity of layers would be even more expensive as more and more locks had to be created, locked and unlocked (all of them syscalls).

course-grained locks vs. fine-grained locks  vs. no locks and do it again!

I had a similar situation in OpenSooq.com. if I had time I'll describe it in a blog post.
in short, we process millions of new photos every day, and generate millions of their previews in various dimensions, in a distributed fashion, and in most case I don't have the original image and I need to fetch it and I had a lock for every image so that it won't be fetched time.
the solution I went with is `not-so-fine-grained locking`, I made lock named after first two hex so at most I would have 256 active locks, there might be a needless lock, but for the one who is having 256 pending downloads then he won't might delaying the 267th one.

**Comment by vrothberg** [comment]
> > consider the costs of locks.
> 
> no matter how many nano seconds the lock cost, it's not as big as downloading openjdk image twice then discard it because it's already downloaded.

The cost of file locks are microseconds (i.e., an oder of magnitude more expensive) with recent CPUs and decent SSDs.

I don't think that the costs of file locks can be compared to the ones of two processes downloading the same data:
- Having such locking will unconditionally slow down *all* copy operations.
- If process A is pulling and B is blocked by a lock, process B does not progress in any case.
- If A and B pull in parallel, chances are that the downloads are faster as the two processes are likely to have different order when downloading blobs.

> > Creating locks on the granularity of layers would be even more expensive as more and more locks had to be created, locked and unlocked (all of them syscalls).
> 
> course-grained locks vs. fine-grained locks vs. no locks and do it again!

I am not sure to understand the message.

> the solution I went with is not-so-fine-grained locking, I made lock named after first two hex so at most I would have 256 active locks, there might be a needless lock, but for the one who is having 256 pending downloads then he won't might delaying the 267th one.

We would need fine-grained locking as coarse-grained locking is not sufficient to avoid redudantly downloading the same layer in parallel as multiple images can share the same base images and hence share a subset of layers.

**Comment by muayyad-alsadi** [comment]
> The cost of file locks are microseconds

still better than downloading openjdk image twice which takes minutes :-) 

> course-grained locks vs. fine-grained locks vs. no locks and do it again!

course-grained in our case is to block as long as other podman process is running (or is running for this image id), one global lock, or one lock per image id.

fine-grained means one lock per layer

no locks and do it again is the current approach seen in the screenshot

> We would need fine-grained locking as coarse-grained locking is not sufficient to avoid redudantly downloading the same layer in parallel as multiple images can share the same base images and hence share a subset of layers.

I'm fine with lock fine-grained lock (per layer) download
but if you want to reduce number of locks you can make it no-too-fine by using the first two hex of layer id.

**Comment by mheon** [comment]
@vrothberg We have a BoltDB cache in c/image now - could we put a pointer in there that a layer is being downloaded elsewhere?

(This has obvious problems if the layer download is interrupted and does not clean up before exiting, so we end up with a bad/invalid state in the DB, but if this could be solved we can potentially avoid complex locking issues)

**Comment by vrothberg** [comment]
> still better than downloading openjdk image twice which takes minutes :-)

It is not. The other process still has to wait for the other pull to complete. Those "minutes" have to be spent waiting in any case. Again, it makes no difference with respect to the time the process must wait for completion. It increases network traffic, which is unfortunate but it is not something I would consider to be a common case.

Assuming we could find a way to implement locking, how would we implement the progress bars? This would imply sending data across processes. I understand that it would be great to have such a feature and docker supports this use case really well (as the progress bars are correctly displayed and simultaneously updated) but I don't see how we could support that.

And again: there is no bug, no data corruption or something similar. We currently cause more network traffic than necessary with the benefit of a deamon-less architecture. And we are talking about something I would consider to be a corner case.

> no locks and do it again is the current approach seen in the screenshot

The terminology is wrong. We don't do it "again" but we do it in parallel. There is a difference between sequentially performing the same tasks or doing it redundantly in parallel. 

> @vrothberg We have a BoltDB cache in c/image now - could we put a pointer in there that a layer is being downloaded elsewhere?

All copying happens in containers/image/copy, so we need to keep track of it there. We also need to consider that any other user of containers/image does it as well.

**Comment by muayyad-alsadi** [comment]
> It is not. The other process still has to wait for the other pull to complete. Those "minutes" have to be spent waiting in any case.

assuming each pull downloads 100MB in 10MB/s network

current behaviour: two downloads, which is total of 200MB and total time of 20 seconds.

desired behaviour: one download, one wait lock release, total for download time is 10 seconds. when lock is released the other process found that it's already downloaded and just skip  in no time.

> And again: there is no bug, no data corruption or something similar.

yes, that's why I put

```
/kind feature
```

> There is a difference between sequentially performing the same tasks or doing it redundantly in parallel.

network throughput/bandwidth is limited. specially if you are pulling external images from outside your local registry.

**Comment by vrothberg** [comment]
> current behaviour: two downloads, which is total of 200MB and total time of 20 seconds.

> 200 MB

Yes, that's the worst case scenario. I think it's unlikely to happen as the layers are unlikely to be downloaded in the same.

> 20 seconds

That would only be true if one process manages to monopolize the entire network I/O and when the registry can provide data at that speed (while excluding the upper case).

I agree that it is not ideal but I don't see how we could support that. The only point of synchronization is github.com/containers/image which is used by Podman, Buildah, Skopeo and CRI-O (and potentially more tools in the wild). Any of those tools can run at the same time. Adding a per-layer copy lock to containers/image entails the issues mentioned in https://github.com/containers/libpod/issues/2551#issuecomment-470079127.

**Comment by mtrmac** [comment]
> > > I don't think we can avoid this "race"
> > 
> > 
> > we can use a lock file, I've done that even in my shell scripts.
> 
> Let me rephrase: I don't think we can avoid this race without major rewrites of the code and we have to consider the costs of locks. The devil lies in the details.
> 
> A big lock as shown in the bashscript is certainly a performance killer, so we had to make it more specific. If we made it specific to the image/ID, we would miss that two images can share the same base image(s). Creating locks on the granularity of layers would be even more expensive as more and more locks had to be created, locked and unlocked (all of them syscalls).

Well, `storageImageDestination.TryReusingBlob` is _already_ locking/unlocking the layer store, for each layer considered (several times!), to check whether the layer exists locally. Checking whether a pull of that layer is in progress is not inconceivable (although we would ideally want to do some more work to only do the locking once per layer.)

[~Unrelated: At a high level, once we have the source manifest, we could lock c/storage _once_ and do all the checks for pre-existing layer presence at the same time, instead of calling `TryReusingLayer` individually and locking in each one fo them. But that would be a more major change to the API.]

`TryReusingBlob` _could_ also, plausibly, wait for the other process to finish pulling, being completely transparent to `c/image/copy`.  Though, that would have to wait for the other process to even do `Commit` so that the layer is actually extracted.  (We’ve been talking about moving the layer extraction out of `Commit` as a possible speed-up anyway, but yes, that would be non-trivial concurrent code, and it’s not strictly necessary for the case we are discussing here.)

I’m _much more_ worried about the cross-process coordination aspect (how does one report progress? What if the process pulling the layer dies?).

> On the other hand, there are more possible destinations than container-storage (e.g., a directory, oci, docker-archive, docker daemon, etc.) and we could also _push_ to all of them. Storing the locks in the storage would hence not suffice as we had to make it transport-generic and specific to the (image, source, destination) triple.

\*shrug\* We can implement this only for c/storage, and leave the current “do the extra work anyway” implementation for the other destinations.  As long as this can be done in `TryReusingBlob`, each transport can be improved independently.

**Comment by mtrmac** [comment]
> @vrothberg We have a BoltDB cache in c/image now - could we put a pointer in there that a layer is being downloaded elsewhere?

- That cache is currently unaware of different c/storage stores and their configurations. That could change, I guess.
- BlotDB is not at all concurrent either - it uses a single file lock over the whole database, held for the whole lifetime of the database handle. So, we end up opening/closing the database, and using a file lock, for every single operation, so that concurrent image copies don’t block on the database for the whole duration of the copy even if they have nothing in common but the cache. That’s only _very slightly_ less expensive than the c/storage locks (no extra path lookup for the lock file, if I understand correctly), and worse for parallelism if c/storage has read-write locks.

**Comment by vrothberg** [comment]
> Well, storageImageDestination.TryReusingBlob is already locking/unlocking the layer store, for each layer considered (several times!), to check whether the layer exists locally.

That's true. In case of the layer store we have 1 persistent lock that protects N layers.

The case here would (potentially) imply N locks for N layers that need to be created (and maybe destroyed) on the fly. Detecting if another process is currenlty copying is tricky as the file locks don't support `tryLock()` semantics; we would be blocked until the lock is acquired. Or we find a clever way doing that in shared memory without leaking data (dying processes) over time.

(going in the code)
*Assuming we had a way to detect if a given layer is currently being copied by another process in a non-blocking fashion*: `TryReusingBlob()` could indicate that via a specific error (e.g., a `CopyInProgressError`) that would cause `copyLayer` to loop.

> I’m much more worried about the cross-process coordination aspect (how does one report progress? What if the process pulling the layer dies?).

Displaying the progress of another process sounds like a herculan task. We could indicate though that the "layer is copied by another process". If another process has died and did not complete the operation, we could query the layer store. If the layer isn't present, we attempt to pull it again. (All assuming we found a way to know coordinate pulling across processes)

**Comment by mtrmac** [comment]
> _Assuming we had a way to detect if a given layer is currently being copied by another process in a non-blocking fashion_: `TryReusingBlob()` could indicate that via a specific error (e.g., a `CopyInProgressError`) that would cause `copyLayer` to loop.

Why would that be useful? `TryReusingBlob` can just as well block until the copy is done and the layer is available at the destination, or until the other process dies and fails (in which case the ordinary `PutBlob` path would be used, exactly what we want) — well, except for progress reporting, which we don’t have for `TryReusingBlob` now.

\#DaemonsAreOccasionallyVeryUseful

**Comment by vrothberg** [comment]
> Why would that be useful? [...] well, except for progress reporting [...]

Yes, for the progress bars. It would be good to show the user what is happening. A workaround would be to update the progress bar with some message prior to acquiring the lock and hoping that we either progress fast enough and update it again before the refresh-rate kicks in or that we wait sufficiently long enough to not have give a flickering impression.

> #DaemonsAreOccasionallyVeryUseful

:hear_no_evil:

**Comment by mtrmac** [comment]
> > Why would that be useful? [...] well, except for progress reporting [...]
> 
> Yes, for the progress bars. It would be good to show the user what is happening.

Still, that could just be passing a “progress update channel” to `TryReusingBlob`, or something similar. (Compare also the ~need to tell the user about the chosen mirror in the c/image/docker mirror PR.)

**Comment by muayyad-alsadi** [comment]
> Yes, for the progress bars. It would be good to show the user what is happening. 

I'll be happy to see a progress in the first one and a message like "layer being pulled/locked by other process" in the other, as when dnf/yum is downloading its metadata

**Comment by vrothberg** [comment]
This is being worked on now. If you interested, you can follow the progress over at https://github.com/containers/skopeo/pull/619. 

Note that I am closing this issue as it's a duplicate of https://github.com/containers/libpod/issues/1911.

Thanks again for taking the time to open and discuss the issue.
